File: packages/hydra-rust/src/hydra/lib/pairs.rs
// hydra.lib.pairs primitives

pub fn first<A: Clone, B>(p: (A, B)) -> A {
    p.0
}

pub fn second<A, B: Clone>(p: (A, B)) -> B {
    p.1
}

pub fn swap<A, B>(p: (A, B)) -> (B, A) {
    (p.1, p.0)
}

pub fn map_first<A, B, C>(f: impl Fn(A) -> C, p: (A, B)) -> (C, B) {
    (f(p.0), p.1)
}

pub fn map_second<A, B, C>(f: impl Fn(B) -> C, p: (A, B)) -> (A, C) {
    (p.0, f(p.1))
}

File: heads/typescript/src/main/typescript/hydra/lib/pairs.ts
// Hand-written runtime: hydra.lib.pairs primitives.

export const first = <A, B>(p: readonly [A, B]): A => p[0];
export const second = <A, B>(p: readonly [A, B]): B => p[1];

export const swap = <A, B>(p: readonly [A, B]): readonly [B, A] => [p[1], p[0]];

export const bimap = <A, B, A2, B2>(f: (a: A) => A2) => (g: (b: B) => B2) => (p: readonly [A, B]): readonly [A2, B2] =>
  [f(p[0]), g(p[1])];
